feat: add support for message.examples

derberg opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow contributing guidelines, the recommended Git workflow, and any related documentation.
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**

Changes proposed in this pull request:

- add support for message.examples
- remove default spec that is not used anywhere
- extended streetlights to have custom examples

Doing serious react and typescript for the first time at the same time is really outside my comfort zone 😅 be gentle
<br><br>
**OLD FUNCTIONALITY STILL WORKS**
<br><br>

<img width="512" alt="Screenshot 2020-11-05 at 13 26 38" src="https://user-images.githubusercontent.com/6995927/98241517-59979480-1f6b-11eb-929d-1abf7108423d.png">
<img width="1099" alt="Screenshot 2020-11-05 at 13 26 30" src="https://user-images.githubusercontent.com/6995927/98241523-5c928500-1f6b-11eb-97eb-f786e435b702.png">

<br><br>
**NEW FUNCTIONALITY LOOKS LIKE THIS**
<br><br>
<img width="592" alt="Screenshot 2020-11-05 at 13 25 21" src="https://user-images.githubusercontent.com/6995927/98241526-5e5c4880-1f6b-11eb-9588-2b9f1ca04054.png">
<img width="1099" alt="Screenshot 2020-11-05 at 13 25 39" src="https://user-images.githubusercontent.com/6995927/98241524-5d2b1b80-1f6b-11eb-8205-5c41e69a0021.png">

**Related issue(s)**
Resolves #92

magicmatatjahu commented on library/src/containers/Schemas/Schema.tsx:
```diff
@@ -52,6 +53,7 @@ export const SchemaComponent: React.FunctionComponent<Props> = ({
   hideTitle = false,
   toggle = false,
   toggleExpand = false,
+  examples,
```
Please default this prop by `examples = []`

magicmatatjahu commented on library/src/containers/Schemas/Schema.tsx:
```diff
@@ -74,8 +76,22 @@ export const SchemaComponent: React.FunctionComponent<Props> = ({
-      {/* we need to disable this component if schema has "not" field anywhere in it */}
-      {hasNotField ? null : (
+
+      {examples ? (
+        (() => {
+          const allExamples: object[] = [];
+          const examplesLength = examples.length;
+          examples.forEach((el, i) => {
+            const title =
+              examplesLength > 1 ? `${exampleTitle} ${i}` : exampleTitle;
+            allExamples.push(
+              <SchemaExampleComponent title={title} example={el} />,
+            );
+          });
+          return allExamples;
+        })()
+      ) : // we need to disable this component if schema has "not" field anywhere in it
```
You can change this lines to simpler code:

```ts
{examples ? examples.map((el, i) => (
   <SchemaExampleComponent title={examples.length > 1 ? `${exampleTitle} ${i}` : exampleTitle} example={el} />,
)) : hasNotField ...}
```

You can also make for `examples.length` additional variable before return statement `const examplesLength = examples.length`.

magicmatatjahu commented on library/src/types.ts:
```diff
@@ -242,10 +242,15 @@ export interface MessageTrait {
 }
 
+export interface Example {
+  headers?: any;
```
`object` type, not any :P

derberg replied:
it was a first change I did when working on this PR, was not yet feeling comfortable with something more than `any` 😄

magicmatatjahu commented on library/src/helpers/getExamplesFromSpec.ts:
```diff
@@ -0,0 +1,14 @@
+
+export type Type = 'payload' | 'headers';
+
+export const getExamplesFromSpec = (examples: Example[], type: Type) => {
```
You can avoid unnecessary forEach and ifs and do something like:

```ts
export const getExamplesFromSpec = (examples: Example[] = [], type: Type) => {
  return examples.map(el => el[type]).filter(Boolean);
}
```

But your solution is also good, my code is only suggestion.

derberg replied:
fuck man, I love it, `filter(Boolean)` is neat!

magicmatatjahu replied:
You're welcome :)

magicmatatjahu reviewed (changes requested):
As in first time in React, it's awesome! And don't lie about TS, you used it in Angular with any type :P

I do reject, because in changed/created functions should be default arguments, other comments (about changing code to simpler) is only suggestion :)

magicmatatjahu commented on library/src/containers/Schemas/SchemaExample.tsx:
```diff
@@ -8,33 +8,35 @@ import { SCHEMA_EXAMPLE_TEXT } from '../../constants';
+    schema && schema.example
+      ? schema.example
+      : schema && generateExampleSchema(schema);
+  const exampleString = JSON.stringify(example || schemaExample, null, 2);
```
Please change it to `JSON.stringify(example || schemaExample || "", null, 2)` because if schema and example props don't exist then pass empty string to JSON.stringify. I know that stringify works also with null or undefined, but this is only for safety.

derberg replied:
yeap, you are right, I changed that schema is also optional now

derberg commented:
@magicmatatjahu type `any` is my favorite 😈 but to be more serious, I actually liked how thanks to TS I could spot basic errors quickly that I would not notic probably with JS and would have to have detailed unit tests

btw, `Simple PR Validation` throws some error related to web component. I guess that your PR should be merged first?

derberg commented:
@magicmatatjahu I'm having some TS headache 😅 

when I change
```
export const getExamplesFromSpec = (examples: Example[], type: Type) => {
  const examplesOfType: object[] = [];
  examples.forEach(el => {
    const ex = el[type];
    if (ex) {
      examplesOfType.push(ex);
    }
  });
  return examplesOfType;
};
```

to
```
export const getExamplesFromSpec = (examples: Example[], type: Type) => {
  examples.map(el => el[type]).filter(Boolean);
}
```

I get some compile error on Message Component. This part 👇 
```
examples={examples && getExamplesFromSpec(examples, 'headers')}
```
screams at me with 👇 
```
Type 'void | undefined' is not assignable to type 'object[] | undefined'.
  Type 'void' is not assignable to type 'object[] | undefined'.ts(2322)
Schema.tsx(23, 3): The expected type comes from property 'examples' which is declared here on type 'IntrinsicAttributes & Props & { children?: ReactNode; 
```

A moment when TS causes issues instead of helping 😛 
Ideas?

magicmatatjahu commented:
@derberg 

> btw, Simple PR Validation throws some error related to web component. I guess that your PR should be merged first?

Yes, at the moment master is broken, so my PR should be merged first

> I get some compile error on Message Component. This part 👇

Please add return statement:

```ts
export const getExamplesFromSpec = (examples: Example[], type: Type) => {
  return examples.map(el => el[type]).filter(Boolean);
}
```

As I see, your body hasn't return.

magicmatatjahu commented on library/src/containers/Schemas/Schema.tsx:
```diff
@@ -68,14 +70,22 @@ export const SchemaComponent: React.FunctionComponent<Props> = ({
-
+  const hasExamples = examples.length;
   const content = (
     <>
       <div className={`${bemClasses.element(`${className}-table`)} p-4`}>
         {renderSchemaProps(name, schema)}
       </div>
-      {/* we need to disable this component if schema has "not" field anywhere in it */}
-      {hasNotField ? null : (
+
+      {hasExamples ? (
+        examples.map((el, i) => (
+          <SchemaExampleComponent
+            title={examples.length > 1 ? `${exampleTitle} ${i}` : exampleTitle}
```
```ts
  const examplesLength = examples.length;
  const content = (
    <>
      <div className={`${bemClasses.element(`${className}-table`)} p-4`}>
        {renderSchemaProps(name, schema)}
      </div>

      {examplesLength ? (
        examples.map((el, i) => (
          <SchemaExampleComponent
            title={examplesLength > 1 ? `${exampleTitle} ${i}` : exampleTitle}
```

magicmatatjahu reviewed (approved):
Very good! Only one suggestion, but it is optional :)

derberg commented:
@magicmatatjahu I updated to the latest master, and after a fresh installation, where I added delay to release pipeline, I also pushed updated to package-lock, have a look, not sure why they appeared.

magicmatatjahu commented:
@derberg Don't worry about package-locks. They are changed, because you use lerna in bootstraping repo, after release they will change to old one with inlined packages from `react-component` :)
